src/evaluators/JSXElement.js

```diff
@@ -221,90 +223,153 @@ function evaluateJSXAttributes(
+            // to mark them as abstract
+          } else {
+            // as the spread might contain "key" and "ref" we have to mark both
+            // as being abstract too, as we don't know if they exist
+            containsAbstractKey = true;
+            containsAbstractRef = true;
+          }
+          // we push the props up to this point into the abstract props args. we also
+          // push the abstract spread object and then we create a fresh props object
+          abstractPropsArgs.push(props, spreadValue);
+          props = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
         }
         break;
       default:
-        invariant(false, `Unknown JSX attribute type:: ${astAttribute.type}`);
+        invariant(false, `Unknown JSX attribute type: ${astAttribute.type}`);
     }
   }
-  return {
-    attributes,
-    children,
-  };
-}
-
-function createReactProps(
-  realm: Realm,
-  type: Value,
-  attributes: Map<string, Value>,
-  children,
-  env: LexicalEnvironment
-): ObjectValue {
-  let obj = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
-  for (let [key, value] of attributes) {
-    if (typeof key === "string") {
-      if (RESERVED_PROPS.hasOwnProperty(key)) {
-        continue;
+  if (containsAbstractSpreadAttribute) {
+    // if we haven't assigned any attributes and we are dealing with a single
+    // spread attribute, we can just make the spread object the props
+    if (attributesAssigned === 0) {
+      props = abstractPropsArgs[1];
+    } else {
+      // we create an abstract Object.assign() to deal with the fact that we don't what
+      // the props are because they contain abstract spread attributes that we can't
+      // evaluate ahead of time
+      let types = new TypesDomain(FunctionValue);
+      let values = new ValuesDomain();
+      let emptyObject = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+      invariant(realm.generator);
+      props = realm.generator.derive(types, values, [emptyObject, ...abstractPropsArgs], _args => {
+        return t.callExpression(
+          t.memberExpression(t.identifier("Object"), t.identifier("assign")),
+          ((_args: any): Array<any>)
+        );
+      });
+      if (
+        spreadAttributesCount === spreadAttributesWithInitialPropsHintCount &&
+        spreadAttributesWithInitialPropsHintCount > 0
+      ) {
+        realm.react.abstractHints.set(props, "HAS_NO_KEY_OR_REF");
+      }
+      if (containsAbstractKey || containsAbstractRef) {
+        // if either are abstract, this will impact the reconcilation process
+        // and ultimately prevent us from folding ReactElements properly
+        // so we unsafely allow this for now, but show a warning
+        invariant(realm.react.logger);
+        realm.react.logger.logWarning(
+          props,
+          `unable to evaluate "key" and "ref" on a ReactElement due to a JSXSpreadAttribute`
+        );
```

Comment:
I don’t feel good about being unsafe by default. I think I would prefer a bailout in this case with a clear message telling me what to fix. If it’s not a hard fix in most cases it could be okay to change the source, and we could make it smarter over time. But this is definitely unsafe.

If you feel strongly about this maybe we could put it behind a flag. `—reactMode unsafe`.

Reply:
Fixed. We could add a flag later for UFI testing.